cmd/kubelet/app/server.go

```diff
@@ -440,6 +445,131 @@ func InitializeTLS(s *options.KubeletServer) (*server.TLSOptions, error) {
 	return tlsOptions, nil
 }
 
+// createCertAndKey tries to:
+// Use the boostrap auth token to get x509 certs if the token is set.
+// If this fails or the token is empty, then self-generates cert and key pair.
+func createCertAndKey(s *options.KubeletServer) error {
+	var err error
+	if s.BootstrapAuthToken != "" {
+		if err = getCertFromAPIServer(s); err != nil {
+			// Clean up cert and key.
+			if err := os.Remove(s.TLSCertFile); err != nil {
+				glog.Warningf("Failed to clean up TLS cert file %q: %v", s.TLSCertFile, err)
+			}
+			if err := os.Remove(s.TLSPrivateKeyFile); err != nil {
+				glog.Warningf("Failed to clean up TLS private key file %q: %v", s.TLSPrivateKey, err)
+			}
+			return fmt.Errorf("Cannot get cert from API server: %v", err)
+		}
+
+		glog.V(4).Infof("Using cert got from the API server (%s, %s)", s.TLSCertFile, s.TLSPrivateKeyFile)
+		return nil
+	}
+
+	if err = crypto.GenerateSelfSignedCert(nodeutil.GetHostname(s.HostnameOverride), s.TLSCertFile, s.TLSPrivateKeyFile, nil, nil); err == nil {
+		glog.V(4).Infof("Using self-signed cert (%s, %s)", s.TLSCertFile, s.TLSPrivateKeyFile)
+		return nil
+	}
+	return fmt.Errorf("unable to generate self-signed cert: %v", err)
+}
+
+// getCertFromAPIServer will:
+// (1) Create a restful client for doing the certificate signing request.
+// (2) Generate key pair and certificate signing request.
+//     The private key is stored to disk.
+// (3) Send request to API server and watch for the issued certificate.
+// (4) Once (3) succeeds, dump the certificate to disk.
+func getCertFromAPIServer(s *options.KubeletServer) error {
+	// (1).
+	clientConfig, err := CreateAPIServerClientConfig(s)
+	if err != nil {
+		return fmt.Errorf("unable to create API server client config: %v", err)
+	}
+	// Embed our bootstrap auth token in each request.
+	clientConfig.BearerToken = s.BootstrapAuthToken
+
+	certificatesclient, err := unversionedcertificates.NewForConfig(clientConfig)
+	if err != nil {
+		return fmt.Errorf("unable to create certificates client: %v", err)
+	}
+
+	// (2).
+	hostname := nodeutil.GetHostname(s.HostnameOverride)
+
+	var ips []net.IP
+	nodeIP := net.ParseIP(s.NodeIP)
+	if nodeIP != nil {
+		ips = []netIP{nodeIP}
+	}
+
+	req, err := utilcertificates.NewCertificateRequest(s.TLSPrivateKeyFile, &pkix.Name{CommonName: "kubelet"}, []string{hostname}, ips)
```

Comment:
The common name is what ends up being treated as the "user" IIRC. I also believe that if you configure `--kubelet-certificate-authority` in the api-server - connections to the node (by nodeName) will validate against the CN.

Maybe we should have common name default to hostname/Nodename instead?